sycl/doc/KernelProgramCache.md

```diff
@@ -68,42 +68,87 @@ times, which may involve JIT compilation and take quite a lot of time.
 
 In order to eliminate this waste of run-time we introduce a kernel and program
 caching. The cache is per-context and it caches underlying objects of non
-interop kernels and programs which are built with no options.
+interop kernels and programs.
+
+Also JIT compilation for cases when an application contains huge amount of
+device code (big kernels or multiple kernels) may take significant time. The
+kernels and programs are rebuild on every program restart. AOT compilation
+can be used to avoid that but it ties application to specific backend runtime
+version(s) and predefined HW configuration(s). As general solution it is
+reasonable to have kernel/program cache which works between application
+restarts (e.g. cache on disk for devices code build for specific HW/SW
+configuration).
 
 <a name="what-is-program">1</a>: Here "program" means an internal SYCL runtime
 object corresponding to a SPIRV module or native binary defining a set of SYCL
 kernels and/or device functions.
 
 
 ## Data structure of cache
+The cache is split into two levels:
+ - in-memory cache which is used during application runtime for device code
+ which has been already loaded and built for target device.
+ - on-disk cache which is used to store device binaries between application
+ executions.
 
+### In-memory cache
 The cache stores underlying PI objects behind `cl::sycl::program` and
 `cl::sycl::kernel` user-level objects in a per-context data storage. The storage
 consists of two maps: one is for programs and the other is for kernels.
 
-The programs map's key consists of three components: kernel set id<sup>[1](#what-is-ksid)</sup>,
-specialized constants, device this program is built for.
+The programs map's key consists of four components: kernel set id<sup>[1](#what-is-ksid)</sup>,
+specialized constants, device this program is built for, build options id <sup>[2](#what-is-bopts)</sup>.
 
-The krnels map's key consists of three components too: program the kernel
-belongs to, kernel name<sup>[2](#what-is-kname)</sup>, device the program is
-built for.
+The kernels map's key consists of four components too: program the kernel
+belongs to, kernel name<sup>[3](#what-is-kname)</sup>, device the program is
+built for, build options id<sup>[2](#what-is-bopts)</sup>.
 
 <a name="what-is-ksid">1</a>: Kernel set id is an ordinal number of the device
 binary image the kernel is contained in.
-
-<a name="what-is-kname">2</a>: Kernel name is a kernel ID mangled class' name
+<a name="what-is-bopts">2</a>: Hash for the string representation of build
+options set in application or environment variables (e.g.
+SYCL_PROGRAM_COMPILE_OPTIONS, SYCL_PROGRAM_LINK_OPTIONS). The key is hash out
```

Comment:
AFAIK there can be options stored in the device image.

Reply:
Options from device image was added to the doc.

Reply:
> Also, matching string hashes does not necessarily mean equivalence strings. Why not use the string itself?

The idea of using hashes instead of fill strings (for build options or device id) to get short representation for using on file system (to avoid path overload issue). There is really small chance of hash collision.

Also build options are goin to be sorted to avoid difference when just order is changed,